Add s390x support for CI scripts

jschintag opened the pull request:
Fixes: #1129

Further Todos:
- build and install qemu will be implemented as soon as the configure hypervisor script supports s390x
- currently the Hot Plug CPU Feature is not working on s390x

jodh-intel commented on .ci/s390x/clean_up_s390x.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+#
+# Copyright (c) 2018 IBM
```
2019 (here and below)?

jschintag replied:
Thank you, i just copied them and forgot to change the year

jodh-intel commented on .ci/s390x/configuration_s390x.yaml:
```diff
@@ -0,0 +1,22 @@
+#
+# Copyright (c) 2018 IBM
+#
+# SPDX-License-Identifier: Apache-2.0#
```
Nit: trailing `#`.

jodh-intel commented on .ci/s390x/lib_install_qemu_s390x.sh:
```diff
@@ -0,0 +1,39 @@
+
+get_packaged_qemu_version() {
+        if [ "$ID" == "ubuntu" ]; then
+		#output redirected to /dev/null
```
I'd drop this comment tbh.

jodh-intel commented on .ci/s390x/lib_install_qemu_s390x.sh:
```diff
@@ -0,0 +1,39 @@
+PACKAGED_QEMU="qemu"
+
+get_packaged_qemu_version() {
+        if [ "$ID" == "ubuntu" ]; then
```
Idea: I'd be tempted to deal with the error scenario first, so in each function:

```sh
[ "$ID" == "ubuntu" ] || die "Unsupported distro: $ID"

 :
```

That will also simplify the remainder of the code as you can remove one layer of `if` testing. Might be worth putting that `die` test at the top global level to avoid having to duplicate either the code in each function, or duplication a call to a function that runs that test?

jschintag replied:
Wouldn't this just mean that i check the content of $ID at two places? I would instead use a switch case to replace this if elif structures and have the error scenario as a default.

jodh-intel replied:
Well, as I say, you could make this a global check in `.ci/s390x/lib_install_qemu_s390x.sh`. Something like:

```sh
[ "$ID" == "ubuntu" ] || die "Unsupported distro: $ID"

get_packaged_qemu_version() {
   # ...
}
```

I think that should work since the caller already sources `/etc/os-release` and `lib.sh` which gives access to `die()`.

jschintag replied:
Yes, it should work at that position. I changed it in the code.

jodh-intel commented on .ci/setup.sh:
```diff
@@ -70,6 +70,16 @@ enable_nested_virtualization() {
+			sudo modprobe -r kvm
+			sudo modprobe kvm nested=1
+			if [ grep -q "N" /sys/module/kvm/parameters/nested 2>/dev/null ]; then
+				die "Failed to find or enable Nested virtualization"
```
This block looks almost identical to `enable_nested_virtualization()` in `.ci/setup.sh`. Could you just update that to handle os390x?

jschintag replied:
sure, i can do that

jodh-intel commented on .ci/s390x/configuration_s390x.yaml:
```diff
@@ -0,0 +1,22 @@
+  - docker
+
+# Same model as is for aarch64
+docker:
```
Comparing these files with the `aarch64` ones, they are *very* similar. I can see that there are small differences but I'm wondering if we can create something like:

```
.ci/generic/*
```

... and have `.ci/aarch64/*` and `.ci/s390x/*` include the generic files if possible. That should work for the shell scripts. I guess it won't work as-is for the YAML as that language doesn't support including natively. But, we could either consider breaking the YAML into chunks and reconstructing or having the scripts use tooling to convert the generic YAML files into arch-specific ones?

Bottom line: it's a maintenance burden on the architecture maintainers to ensure multiple very similar files remain updated. But maybe these files won't change much?

alicefr replied:
hi @jodh-intel now the config looks pretty similar to arm, but on the other hand this file lists (or better black list) what's not working/have been tested on s390x. It could be that in the future the 2 architectures files diverge. On s390x for example, qemu hasn't supported memory hotplug  yet. On s390x, the CPU hotplug has probably a bug with Kata but it should be supported. The file could be a bit redundant but it clearly list what has been tested on s390x. Otherwise, we can have a discussion with @Pennyzct and @nitkon how to integrate the tested features in a more reasonable way for all the arch :)

jodh-intel replied:
Hi @alicefr - Yes, let's leave these files as separate arch-specific ones for now. We can always re-assess the situation again in the future.

grahamwhaley commented on .ci/s390x/filter_docker_s390x.sh:
```diff
@@ -0,0 +1,51 @@
+	done
+}
+
+main()
```
Similar to some other items, much of this filter code looks arch generic - at some point we should probably pull a bunch of this out into library functions.
Does not have to block this PR, but will /cc @nitkon @GabyCT on it.

grahamwhaley commented on .ci/setup.sh:
```diff
@@ -57,11 +57,16 @@ install_docker() {
 	case "$arch" in
-	x86_64)
+	x86_64 | s390x)
 		if [ "$CI" == true ] && grep -q "N" /sys/module/kvm_intel/parameters/nested 2>/dev/null; then
```
given you are doing a non-intel modprobe below, is this `kvm_intel` enabled check also not valid for s390x ?

jschintag replied:
Yes it isn't. I overlooked that one.

grahamwhaley commented on integration/docker/docker.go:
```diff
@@ -34,10 +34,7 @@ const (
-
-	// CentosImage is the centos image
-	CentosImage = "centos"
+        FedoraImage = "fedora"
```
Is this an intentional change - unclear to me (might have been clearer if it was in its own commit ;-) )

grahamwhaley commented on integration/docker/docker_arch_base.go:
```diff
@@ -0,0 +1,6 @@
+//+build !s390x
```
I don't think we generally use in-file golang build constraints in kata - the constraint via the filename _extention is our normal way. wdty @jodh-intel ?

jodh-intel replied:
Yes, but you can't negate using the filename afaik, so this one is fine.

grahamwhaley commented:
/test
coz I want to see how it flies....

chavafg commented:
I see `.ci/.install_qemu.sh.swp`, which shouldn't be part of this change

chavafg commented on .ci/s390x/lib_install_qemu_s390x.sh:
```diff
@@ -0,0 +1,36 @@
+
+set -e
+
+CURRENT_QEMU_VERSION=$(get_version "assets.hypervisor.qemu.version")
```
seems like this is not being used, so it can be removed until we use it for building qemu.

jschintag commented on .ci/install_qemu.sh:
```diff
@@ -91,25 +91,34 @@ if [ -f "${cidir}/${QEMU_ARCH}/lib_install_qemu_${QEMU_ARCH}.sh" ]; then
+			fi
+			;;
+		"s390x")
+			install_packaged_qemu
```
No, it will not build the source. As i've written in the PR-Message, building Qemu from source is currently not implemented. It will be added later, as soon as the configure_hypervisor.sh script from packaging supports s390x. The code will then look similar to x86_64.

alicefr replied:
@grahamwhaley if this is ok for you we can add the qemu building later. Otherwise, this PR is blocked by the missing support in [configure-hypervisor.sh](https://github.com/kata-containers/packaging/blob/master/scripts/configure-hypervisor.sh). As we obtained the CI integrated, we thought to open the pr to be able at least to install the dependencies and build Kata.

grahamwhaley replied:
Sure, its fine - was just checking...

jschintag commented:
I added the feedback from @grahamwhaley and @chavafg

alicefr commented on .ci/setup.sh:
```diff
@@ -57,12 +57,15 @@ install_docker() {
+	x86_64 | s390x)
+		kvm_arch="kvm"
+		[ ${arch} == "x86_64" ] && kvm_arch="kvm_intel"
+		echo "${kvm_arch}"
```
Do we need this print?

jschintag replied:
no we don't

grahamwhaley commented:
Hi @jschintag - thanks for the updates. We are getting there :-)
The CI check is moaning that you have no text body to your commit though - can you add a brief description (or, as full as you like ;-) ) pls.

```
Found 1 commit between commit 5ea2931de74c090c3246ab52db64bcc59e32177f and branch master
ERROR: Commit 5ea2931de74c090c3246ab52db64bcc59e32177f: single-line "Signed-off-by" body not permitted
ERROR: checkcommits failed. See the document below for help on formatting
commits for the project.
https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format
```

grahamwhaley commented:
/test

jodh-intel commented on integration/docker/docker_s390x.go:
```diff
@@ -0,0 +1,8 @@
+//+build s390x
```
Nit: not required as you've already encoded the platform in the name of the file itself.

jschintag commented:
@grahamwhaley  Ok, i added a body to the commit message. Hopefully the CI check likes it now.
@jodh-intel I also removed the build constraint header in integration/docker/docker_s390x.go

grahamwhaley commented:
Hi @jschintag - CI happy with the commit now, but it has noticed two files don't have the SPDX copyright headers...

```
All commit checks passed.
INFO: Checking for SPDX license headers
ERROR: Required license identifier ('SPDX-License-Identifier: Apache-2.0') missing from following files:
integration/docker/docker_arch_base.go
integration/docker/docker_s390x.go
The command "bash .ci/static-checks.sh github.com/kata-containers/tests" exited with 1.
```

I know, battling the CIs... but, I'm hoping we are nearly there!

jschintag commented:
@grahamwhaley No problem, i'll fix all the Problems the CI has.

grahamwhaley commented:
side note - @ttx, look, Zuul SoB/WIP checks active and working on this tests repo PR :-)

grahamwhaley commented:
@jschintag thanks - Travis now happy, along with the new Zuul checks (that replace some of the broken pullapprove CI). I'll fire the tests again so we can check we have not broken any other arch's...

grahamwhaley commented:
/test

grahamwhaley commented:
From the 16.04 CI, looks like we have some code issue:

```
14:39:05 bash sanity/check_sanity.sh
14:39:05 ./ginkgo -v -focus "" -skip "" ./integration/docker/ -- -runtime=kata-runtime -timeout=60
14:39:05 Failed to compile docker:
14:39:05 
14:39:05 can't load package: package github.com/kata-containers/tests/integration/docker: 
14:39:05 integration/docker/docker_arch_base.go:8:1: expected 'package', found 'const'
14:39:05 integration/docker/docker_arch_base.go:10:9: expected ';', found CentosImage
```

jschintag commented:
I fixed the issue

jodh-intel commented:
Thanks @jschintag.

/retest

jschintag commented:
@grahamwhaley @jodh-intel Since https://github.com/kata-containers/packaging/pull/333 was merged i added the support for building qemu on s390x.

grahamwhaley commented on integration/docker/docker_arch_base.go:
```diff
@@ -0,0 +1,12 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+package docker
```
odd, I think the CI is moaning about this, but did not moan about the one in docker_s390x.go?
```
integration/docker/docker_arch_base.go:3:1:warning: package comment should be of the form "Package docker ..." (golint)
The command "bash .ci/static-checks.sh github.com/kata-containers/tests" exited with 1.
```

@sameo - is this maybe at all related to the static check stuff you were looking at yesterday? (that is, not that you broke it, but maybe you know if there is something related to what you were debugging/fixing etc.).

jschintag replied:
It seems the CI wants at least one empty line between the package and the License

chavafg commented:
/test

jodh-intel commented:
@chavafg - the firecracker CI is failing but that doesn't appear to be related to this PR. Can we just merge this as all other CI's are passing?

chavafg commented:
sure, lets merge, but I think this needs 1 more ack

grahamwhaley commented:
Only fc failing - let's merge... and, let's figure out what the deal with fc is...
